group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -3898,6 +3994,65 @@ public CoordinatorResult<Void, Record> classicGroupSync(
         return EMPTY_RESULT;
     }
 
+    /**
+     * Handle a SyncGroupRequest to a ConsumerGroup.
+     *
+     * @param group          The ConsumerGroup.
+     * @param context        The request context.
+     * @param request        The actual SyncGroup request.
+     * @param responseFuture The sync group response future.
+     *
+     * @return The result that contains records to append.
+     */
+    private CoordinatorResult<Void, Record> classicGroupSyncToConsumerGroup(
+        ConsumerGroup group,
+        RequestContext context,
+        SyncGroupRequestData request,
+        CompletableFuture<SyncGroupResponseData> responseFuture
+    ) throws UnknownMemberIdException, GroupIdNotFoundException {
+        String groupId = request.groupId();
+        String memberId = request.memberId();
+        String instanceId = request.groupInstanceId();
+
+        ConsumerGroupMember member;
+        if (instanceId == null) {
+            member = group.getOrMaybeCreateMember(request.memberId(), false);
+        } else {
+            member = group.staticMember(instanceId);
+            if (member == null) {
+                throw new UnknownMemberIdException(
+                    String.format("Member with instance id %s is not a member of group %s.", instanceId, groupId)
+                );
+            }
+            throwIfInstanceIdIsFenced(member, groupId, memberId, instanceId);
+        }
+
+        throwIfMemberDoesNotUseClassicProtocol(member);
+        throwIfGenerationIdOrProtocolUnmatched(
+            group,
+            member,
+            request.generationId(),
+            request.protocolType(),
+            request.protocolName()
+        );
+
```

Comment:
I wanted to make it return REBALANCE_IN_PROGRESS but we seems to have no way to know whether a new rebalance is triggered in sync group...

Reply:
Hum... I wonder if we could rely on the member epoch to do this. If the member epoch is smaller than the group epoch and the member is not in unrevoked partitions state, we could return rebalance in progress. If the member epoch is smaller than the group epoch, it means that the member must rebalance to catch up. However, if the member is already in unrevoked partitions state, it means that it has already started a rebalance and it must complete it to revoke partitions. It will automatically start another one after revoking the partitions. Would something like this work?

Reply:
Yeah! I was stuck in how to determine if the member has already rejoined but checking the state makes sense